File: ios/SkyCast/Core/DesignSystem/WeatherDetailGrid.swift
import SwiftUI

/// One labelled reading, e.g. "Humidity" / "69%".
///
/// `value` is an already-formatted display string: how many decimal places, which unit symbol and
/// what time format are presentation decisions, so they happen in the feature layer
/// (`Features/Common/WeatherDetails.swift`) and the design system just renders what it is given.
///
/// The Android counterpart is `core/designsystem/component/WeatherDetail.kt`.
struct WeatherDetail: Identifiable, Equatable {
    let label: String
    let value: String
    let kind: WeatherDetailKind
    /// How the reading draws itself. See ``MetricVisual`` for why this is not simply a fraction.
    var visual: MetricVisual = .plain

    var id: String {
        label
    }
}

/// Which reading a tile shows.
///
/// An enum rather than a colour or symbol passed in by the caller, so the design system owns how
/// each metric looks and a feature cannot give humidity two different tints on two screens.
enum WeatherDetailKind: CaseIterable {
    case humidity
    case wind
    case pressure
    case visibility
    case sunrise
    case sunset

    /// Reported all along and never shown until the tiles could draw it.
    case cloud

    // Derived rather than reported, see ``Weather/dewPointCelsius`` and
    // ``Weather/daylightDuration``. They share the hue of the reading they are closest to: dew
    // point is moisture, so it takes humidity's blue, and daylight takes the sunrise gold.
    case dewPoint
    case daylight

    var symbolName: String {
        switch self {
        case .humidity: "humidity.fill"
        case .wind: "wind"
        case .pressure: "barometer"
        case .visibility: "eye.fill"
        case .sunrise: "sunrise.fill"
        case .sunset: "sunset.fill"
        case .cloud: "cloud.fill"
        case .dewPoint: "thermometer.medium"
        case .daylight: "sun.horizon.fill"
        }
    }

    var accent: Color {
        switch self {
        case .humidity, .dewPoint: WeatherPalette.humidity
        case .wind: WeatherPalette.wind
        case .pressure: WeatherPalette.pressure
        case .visibility: WeatherPalette.visibility
        case .sunrise, .daylight: WeatherPalette.sunrise
        case .sunset: WeatherPalette.sunset
        case .cloud: WeatherPalette.onCloudContainer
        }
    }
}

/// The secondary readings: humidity, wind, pressure, visibility, sunrise, sunset.
///
/// A `LazyVGrid` with an **adaptive** column, so the tiles reflow instead of clipping when the user
/// turns Dynamic Type up. A fixed two-column grid breaks at accessibility sizes.
///
/// Each tile carries its metric's colour and, where the reading has a scale, an indicator showing
/// where on that scale it sits.
///
/// No glass: these tiles sit inside a page of content rather than floating over it.
struct WeatherDetailGrid: View {
    let details: [WeatherDetail]

    var body: some View {
        LazyVGrid(
            columns: [GridItem(.adaptive(minimum: minimumTileWidth), spacing: Spacing.sm)],
            spacing: Spacing.sm
        ) {
            ForEach(details) { detail in
                WeatherDetailTile(detail: detail)
            }
        }
    }

    /// Wide enough for "1009 hPa" at the default text size; the adaptive column then decides how
    /// many fit, dropping to one per row at accessibility sizes.
    private let minimumTileWidth: CGFloat = 150
}

private struct WeatherDetailTile: View {
    let detail: WeatherDetail

    /// The tint of the page this tile is on, if it has one.
    @Environment(\.weatherSurfaceTint) private var weatherSurfaceTint

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.xs) {
            Label {
                Text(detail.label)
                    .font(.caption)
                    .foregroundStyle(.secondary)
            } icon: {
                Image(systemName: detail.kind.symbolName)
                    .font(.caption)
                    .foregroundStyle(detail.kind.accent)
            }

            HStack(alignment: .center, spacing: Spacing.sm) {
                Text(detail.value)
                    .font(.title3.weight(.medium))
                    .minimumScaleFactor(0.7)
                    .lineLimit(1)

                Spacer(minLength: 0)

                switch detail.visual {
                case .plain:
                    EmptyView()
                case let .gauge(fraction):
                    MetricGauge(fraction: fraction, colour: detail.kind.accent)
                case let .compass(degrees, _):
                    WindCompass(degrees: degrees, colour: detail.kind.accent)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(Spacing.md)
        // The surface first, then a whisper of the page's mood over it. Layered rather than
        // blended into a single colour so the base stays the one the system guarantees contrast
        // against, the tile belongs to a warm page or a cold one without any text on it becoming
        // a contrast problem that has to be re-checked per condition.
        .background {
            RoundedRectangle(cornerRadius: Radius.md, style: .continuous)
                .fill(Color.skySurface)
                .overlay {
                    if let weatherSurfaceTint {
                        RoundedRectangle(cornerRadius: Radius.md, style: .continuous)
                            .fill(weatherSurfaceTint.opacity(tintOpacity))
                    }
                }
        }
        // One announcement per tile. Without this, VoiceOver reads "Humidity" and "69%" as two
        // unrelated fragments and the pairing is lost. The bar is decorative, it says the same
        // thing as the value.
        .accessibilityElement(children: .ignore)
        .accessibilityLabel(announcement)
    }

    /// The compass draws a bearing that the value alone does not mention, so it is spoken.
    private var announcement: String {
        if case let .compass(_, cardinal) = detail.visual {
            return "\(detail.label), \(detail.value) from the \(cardinal)"
        }
        return "\(detail.label), \(detail.value)"
    }

    /// Enough to be felt when swiping between a clear place and an overcast one, little enough
    /// that the tile still reads as a surface rather than a coloured chip.
    ///
    /// Higher than it was, and lighter for it: the previous 10% of the *mood* hue put a mid-dark
    /// slate over a white card, and the result read as dirty grey beside the near-white glass next
    /// to it. 45% of the container keeps a light card light and still carries the condition.
    private let tintOpacity: Double = 0.45
}

#Preview {
    ScrollView {
        WeatherDetailGrid(details: [
            WeatherDetail(label: "Humidity", value: "69%", kind: .humidity, visual: .gauge(0.69)),
            WeatherDetail(
                label: "Wind",
                value: "8.7 kt",
                kind: .wind,
                visual: .compass(degrees: 270, cardinal: "W")
            ),
            WeatherDetail(label: "Pressure", value: "29.92 inHg", kind: .pressure, visual: .gauge(0.45)),
            WeatherDetail(label: "Visibility", value: "5.4 NM", kind: .visibility, visual: .gauge(1)),
            WeatherDetail(label: "Cloud cover", value: "40%", kind: .cloud, visual: .gauge(0.4)),
            WeatherDetail(label: "Dew point", value: "12°C", kind: .dewPoint, visual: .gauge(0.6)),
        ])
        .padding(Spacing.md)
    }
    .background(Color.skyBackground)
}

File: android/core/designsystem/src/main/kotlin/com/nauhaan/skycast/core/designsystem/component/WeatherDetailGrid.kt
package com.nauhaan.skycast.core.designsystem.component

import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Air
import androidx.compose.material.icons.filled.Compress
import androidx.compose.material.icons.filled.DeviceThermostat
import androidx.compose.material.icons.filled.LightMode
import androidx.compose.material.icons.filled.Visibility
import androidx.compose.material.icons.filled.WaterDrop
import androidx.compose.material.icons.filled.WbSunny
import androidx.compose.material.icons.filled.WbTwilight
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.compositeOver
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.semantics.clearAndSetSemantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.nauhaan.skycast.core.designsystem.theme.LocalWeatherTint
import com.nauhaan.skycast.core.designsystem.theme.SkyCastTheme
import com.nauhaan.skycast.core.designsystem.theme.Spacing
import com.nauhaan.skycast.core.designsystem.theme.weatherPalette

/**
 * The secondary readings, humidity, wind, pressure, visibility, sunrise, sunset.
 *
 * A `FlowRow` rather than a fixed grid so the tiles reflow instead of clipping at large font
 * sizes, which is what the Dynamic Type requirement actually demands.
 *
 * Each tile carries its own colour and, where the reading has a scale, an indicator showing where
 * on that scale it sits. "1014 hPa" tells most people nothing; a bar sitting just past the middle
 * tells them it is an ordinary day. The colours come from the weather palette, so humidity is the
 * same blue everywhere it appears.
 *
 * Each tile announces as one unit; without that, TalkBack reads a label and a bare number as two
 * unrelated fragments.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun WeatherDetailGrid(details: List<WeatherDetail>, modifier: Modifier = Modifier) {
    FlowRow(
        modifier = modifier.fillMaxWidth(),
        horizontalArrangement = Arrangement.spacedBy(Spacing.sm),
        verticalArrangement = Arrangement.spacedBy(Spacing.sm),
        maxItemsInEachRow = TILES_PER_ROW,
    ) {
        details.forEach { detail ->
            WeatherDetailTile(
                detail = detail,
                // weight(1f) inside a FlowRow gives equal-width tiles that still reflow: at large
                // font sizes a row simply holds fewer of them.
                modifier = Modifier.weight(1f),
            )
        }
    }
}

@Composable
private fun WeatherDetailTile(detail: WeatherDetail, modifier: Modifier = Modifier) {
    val accent = detail.kind.accent()

    // The theme surface with a whisper of the page's mood mixed over it. Composited rather than
    // replaced so the base stays the colour the theme guarantees contrast against: the tile belongs
    // to a warm page or a cold one without any text on it becoming a contrast problem that has to
    // be re-checked per condition. Animated on the effects spec, this is a colour change, and the
    // spatial spec would overshoot past the target colour.
    val tint = LocalWeatherTint.current
    val container by animateColorAsState(
        targetValue = MaterialTheme.colorScheme.surfaceContainerHigh.let { surface ->
            if (tint == null) surface else tint.copy(alpha = TINT_ALPHA).compositeOver(surface)
        },
        animationSpec = MaterialTheme.motionScheme.defaultEffectsSpec(),
        label = "detailTileContainer",
    )

    Card(
        modifier = modifier.clearAndSetSemantics {
            contentDescription = "${detail.label}, ${detail.value}"
        },
        colors = CardDefaults.cardColors(containerColor = container),
    ) {
        Column(modifier = Modifier.padding(Spacing.md)) {
            Row(verticalAlignment = Alignment.CenterVertically) {
                Icon(
                    imageVector = detail.kind.icon(),
                    // Decorative: the label beside it says the same thing in words.
                    contentDescription = null,
                    tint = accent,
                    modifier = Modifier.size(IconSize),
                )
                Text(
                    text = detail.label,
                    style = MaterialTheme.typography.labelMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                    modifier = Modifier.padding(start = Spacing.xs),
                )
            }
            Text(
                text = detail.value,
                style = MaterialTheme.typography.titleMedium,
                modifier = Modifier.padding(top = Spacing.xs),
            )
            detail.fraction?.let { fraction ->
                MetricBar(
                    fraction = fraction.toFloat(),
                    colour = accent,
                    modifier = Modifier.padding(top = Spacing.sm),
                )
            }
        }
    }
}

/**
 * A slim bar showing where a reading sits on its scale.
 *
 * Animated with the Expressive **effects** spec, not the spatial one: this is a colour-and-length
 * change, and a spatial spec would overshoot past the value before settling, a bar that briefly
 * shows 105% humidity. That distinction is the rule most often broken.
 */
@Composable
private fun MetricBar(fraction: Float, colour: Color, modifier: Modifier = Modifier) {
    val animated by animateFloatAsState(
        targetValue = fraction.coerceIn(0f, 1f),
        animationSpec = MaterialTheme.motionScheme.defaultEffectsSpec(),
        label = "metricBar",
    )

    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(BarHeight)
            .clip(MaterialTheme.shapes.extraSmall)
            .background(weatherPalette.metricTrack)
            // Drawn rather than laid out. A child `Layout` sized to a fraction of `maxWidth`
            // crashed the app: `FlowRow` with weights measures intrinsics, during which
            // `maxWidth` is infinite, and `Infinity * fraction` overflowed the size limit.
            // `drawBehind` runs after measurement with a real size, so intrinsics cannot reach it.
            .drawBehind {
                drawRect(color = colour, size = Size(size.width * animated, size.height))
            },
    )
}

private fun WeatherDetailKind.icon(): ImageVector = when (this) {
    WeatherDetailKind.HUMIDITY -> Icons.Filled.WaterDrop
    WeatherDetailKind.WIND -> Icons.Filled.Air
    WeatherDetailKind.PRESSURE -> Icons.Filled.Compress
    WeatherDetailKind.VISIBILITY -> Icons.Filled.Visibility
    WeatherDetailKind.SUNRISE -> Icons.Filled.WbSunny
    WeatherDetailKind.SUNSET -> Icons.Filled.WbTwilight
    WeatherDetailKind.DEW_POINT -> Icons.Filled.DeviceThermostat
    WeatherDetailKind.DAYLIGHT -> Icons.Filled.LightMode
}

@Composable
private fun WeatherDetailKind.accent(): Color = when (this) {
    WeatherDetailKind.HUMIDITY -> weatherPalette.humidity
    WeatherDetailKind.WIND -> weatherPalette.wind
    WeatherDetailKind.PRESSURE -> weatherPalette.pressure
    WeatherDetailKind.VISIBILITY -> weatherPalette.visibility
    WeatherDetailKind.SUNRISE -> weatherPalette.sunrise
    WeatherDetailKind.SUNSET -> weatherPalette.sunset
    WeatherDetailKind.DEW_POINT -> weatherPalette.humidity
    WeatherDetailKind.DAYLIGHT -> weatherPalette.sunrise
}

/** Two tiles per row at the default text size; FlowRow drops to one when they stop fitting. */
private const val TILES_PER_ROW = 2

/**
 * Enough to be felt when swiping between a clear place and an overcast one, little enough that the
 * tile still reads as a neutral surface rather than a coloured chip.
 */
private const val TINT_ALPHA = 0.10f
private val IconSize = 16.dp
private val BarHeight = 4.dp

@Preview(showBackground = true)
@Composable
private fun WeatherDetailGridPreview() {
    SkyCastTheme {
        WeatherDetailGrid(
            details = listOf(
                WeatherDetail("Humidity", "69%", WeatherDetailKind.HUMIDITY, fraction = 0.69),
                WeatherDetail("Wind", "4.5 m/s", WeatherDetailKind.WIND, fraction = 0.18),
                WeatherDetail("Pressure", "1009 hPa", WeatherDetailKind.PRESSURE, fraction = 0.45),
                WeatherDetail("Visibility", "10.0 km", WeatherDetailKind.VISIBILITY, fraction = 1.0),
                WeatherDetail("Sunrise", "05:27", WeatherDetailKind.SUNRISE),
                WeatherDetail("Sunset", "20:46", WeatherDetailKind.SUNSET),
            ),
            modifier = Modifier.padding(Spacing.md),
        )
    }
}
